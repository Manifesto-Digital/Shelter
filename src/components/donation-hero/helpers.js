export const formatDonationContent = content => {
  const createOptionObject = (amount, description, image) => ({
    amount,
    description,
    image,
  });

  // single
  const { singleAmount1, singleDescription1, singleImage1 } = content;
  const singleOption1 = createOptionObject(
    singleAmount1,
    singleDescription1,
    singleImage1
  );
  const { singleAmount2, singleDescription2, singleImage2 } = content;
  const singleOption2 = createOptionObject(
    singleAmount2,
    singleDescription2,
    singleImage2
  );
  const { singleAmount3, singleDescription3, singleImage3 } = content;
  const singleOption3 = createOptionObject(
    singleAmount3,
    singleDescription3,
    singleImage3
  );
  const { singleAmount4, singleDescription4, singleImage4 } = content;
  const singleOption4 = createOptionObject(
    singleAmount4,
    singleDescription4,
    singleImage4
  );

  const { singleAmount5, singleDescription5, singleImage5 } = content;
  const singleOption5 = createOptionObject(
    singleAmount5,
    singleDescription5,
    singleImage5
  );

  // monthly
  const { monthlyAmount1, monthlyDescription1, monthlyImage1 } = content;
  const monthlyOption1 = createOptionObject(
    monthlyAmount1,
    monthlyDescription1,
    monthlyImage1
  );
  const { monthlyAmount2, monthlyDescription2, monthlyImage2 } = content;
  const monthlyOption2 = createOptionObject(
    monthlyAmount2,
    monthlyDescription2,
    monthlyImage2
  );
  const { monthlyAmount3, monthlyDescription3, monthlyImage3 } = content;
  const monthlyOption3 = createOptionObject(
    monthlyAmount3,
    monthlyDescription3,
    monthlyImage3
  );
  const { monthlyAmount4, monthlyDescription4, monthlyImage4 } = content;
  const monthlyOption4 = createOptionObject(
    monthlyAmount4,
    monthlyDescription4,
    monthlyImage4
  );
  return {
    single: [
      singleOption1,
      singleOption2,
      singleOption3,
      singleOption4,
      singleOption5,
    ],
    monthly: [monthlyOption1, monthlyOption2, monthlyOption3, monthlyOption4],
  };
};
