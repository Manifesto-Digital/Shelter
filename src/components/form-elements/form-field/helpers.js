const possibleTextInputs = Object.freeze([
  'email',
  'text-field',
  'phone-number',
  'numeric',
  'text-area',
]);

export const isTextField = fieldType => possibleTextInputs.includes(fieldType);

export const getTextInputType = fieldType => {
  if (fieldType === 'phone-number') return 'tel';
  if (fieldType === 'numeric') return 'number';
  if (fieldType === 'email') return fieldType;
  if (fieldType === 'text-area') return 'textarea';
  return 'text';
};
