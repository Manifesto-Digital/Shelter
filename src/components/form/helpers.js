import * as Yup from 'yup';
import { consistentString } from '../../utils/content-formatting';

const isMultipleCheckbox = field =>
  consistentString(field.fieldType) === 'checkboxes' &&
  field.defaultValue &&
  field.valueOptions.length > 1;

export const getInitialValues = (fields, hiddenInitialValues) => {
  const initialValues = hiddenInitialValues;

  // Loop through all fields to build an object to store initial values to pass to formik
  fields.forEach(field => {
    if (!field.fieldType) return;

    if (field.fieldType === 'Address') return; // Temporary

    // Set default value to be the initial value for form
    if (field.internal.type === 'ContentfulTopicFormField') {
      initialValues[field.machineName] = field.defaultValue || '';
    }
    // Multiple checkboxes require an array of values
    if (isMultipleCheckbox(field)) {
      initialValues[field.machineName] = field.defaultValue
        ? [field.defaultValue]
        : [];
    }

    if (field.internal.type === 'ContentfulTopicFormFieldset') {
      getInitialValues(field.formFields);
    }
  });
  return initialValues;
};

export const getValidationSchema = formFields => {
  const validationSchema = {};

  // Loop through all form fields and build a Yup validationSchema object for formik
  formFields.forEach(field => {
    if (!field.fieldType) return;
    const fieldType = consistentString(field.fieldType);
    if (field.fieldType === 'address') return; // Temporary
    const fieldName = field.machineName;

    let validationType;
    // Apply a validation type to each field
    if (isMultipleCheckbox(field)) validationType = Yup.array();
    else if (fieldType === 'phone-number') {
      validationType = Yup.number().typeError('Please provide a valid number');
    } else if (fieldType === 'email') {
      validationType = Yup.string().email('Please provide a valid email');
    } else validationType = Yup.string();

    // Add required validation if specified
    if (field.required) validationType = validationType.required('Required');

    validationSchema[fieldName] = validationType;
  });

  return Yup.object().shape(validationSchema);
};
