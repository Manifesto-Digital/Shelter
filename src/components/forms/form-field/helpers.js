const possibleTextInputs = Object.freeze([
  'email',
  'text-field',
  'phone-number',
  'numeric',
]);

export const isTextField = fieldType => possibleTextInputs.includes(fieldType);

export const getTextInputType = fieldType => {
  if (fieldType === 'phone-number') return 'tel';
  if (fieldType === 'numeric') return 'number';
  if (fieldType === 'email') return fieldType;
  return 'text';
};
