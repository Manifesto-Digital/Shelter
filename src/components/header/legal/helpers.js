/**
 *
 * Loop through the professionals navigation menu to find the current active page
 * and the parent pages so to that the correct menus and active states will be shown
 *
 * @param {Object []} childMenuItems
 * @param {Object} location
 *
 * @returns [menuParentSlug, menuParentSlug, activePageSlug] (example)
 */
export const getActivePages = (childMenuItems, location) => {
  const isActivePage = slug => {
    // Match optional trailing slash
    const string = `${slug}((/w+)+|/?)$`;
    const regex = new RegExp(string, 'g');
    return location.pathname.match(regex);
  };

  let activePagesArray = [];
  let pageFound = false;

  const checkLevel = items => {
    for (let i = 0; i < items.length; i += 1) {
      const item = items[i];
      if (!item.navigationLink) return;

      const itemSlug = item.navigationLink[0].slug;

      if (isActivePage(itemSlug)) {
        // If found push the active page slug
        activePagesArray.push(itemSlug);
        pageFound = true;
        break;
      } else if (item.childNavigationItems) {
        // Recursively go through the child navigation items storing
        // potential active parents item slugs
        activePagesArray.push(itemSlug);
        checkLevel(item.childNavigationItems);
      }
    }
    if (!pageFound) activePagesArray = [];
    return activePagesArray;
  };

  checkLevel(childMenuItems);
  return activePagesArray;
};
