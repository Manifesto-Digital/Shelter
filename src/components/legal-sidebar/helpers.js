export const buildCurrentPageHierarchy = (hierarchy, parentSlug) => {
  const hierarchyDepth = 2;

  // Get the parent of the topmost level to show for use in the heading.
  const headingRef =
    hierarchyDepth > parentSlug.length
      ? parentSlug[parentSlug.length - hierarchyDepth - 1]
      : parentSlug[0];
  const heading = headingRef.label;

  // The top most level is Legal which we do not want to show so remove from parentSlug
  const legalChildrenSlugs = parentSlug.slice(1);

  // Slice the pages parent slug if the current page greater than hierarchy depth
  const topLevelToShowIndex = legalChildrenSlugs.length - hierarchyDepth;
  const parentHierarchyToShow =
    legalChildrenSlugs.length > hierarchyDepth
      ? legalChildrenSlugs.slice(topLevelToShowIndex)
      : legalChildrenSlugs;

  // Check if page exists on current hierarchy level
  function existsOnLevel(pageRef) {
    return parentHierarchyToShow.find(ref => ref.slug === pageRef.slug);
  }

  // Copy to avoid mutating original hierarchy
  const hierarchyCopy = JSON.parse(JSON.stringify(hierarchy));

  // Loop through assign an active state to the current page parents
  function buildActiveHierarchy(hierarchyLevel) {
    const newHierarchy = [];
    function recursiveLoop(level) {
      level.forEach((pageRef, i) => {
        if (existsOnLevel(pageRef)) {
          // Set current item to active
          level[i].active = true;
          // Push the first time this matches, this will also include the children
          if (newHierarchy.length === 0) newHierarchy.push(level);
        }
        if (pageRef.children) {
          recursiveLoop(pageRef.children);
        }
      });
    }
    recursiveLoop(hierarchyLevel);

    return newHierarchy;
  }

  const currentPageHierarchy = buildActiveHierarchy(hierarchyCopy);
  return { currentPageHierarchy: currentPageHierarchy[0], heading };
};
