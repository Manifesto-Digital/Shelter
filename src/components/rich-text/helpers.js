export const fieldsMap = fieldsToMap =>
  Object.keys(fieldsToMap).reduce((accum, key) => {
    const field = fieldsToMap[key];
    if (Object.prototype.hasOwnProperty.call(field, 'en-GB')) {
      accum[key] = flattenKey('en-GB', field); // eslint-disable-line no-use-before-define
    } else if (typeof field === 'object' && field !== null) {
      accum[key] = fieldsMap(field);
    } else {
      accum[key] = field;
    }
    return accum;
  }, {});

const flattenKey = (key, field) => {
  if (typeof field[key] === 'string' || typeof field[key] === 'boolean') {
    return field[key];
  }
  if (Array.isArray(field[key]) && field[key].length === 1) {
    return [fieldsMap(field[key][0].fields)];
  }
  if (typeof field[key] === 'object') {
    if (Object.prototype.hasOwnProperty.call(field[key], 'fields')) {
      return fieldsMap(field[key].fields);
    }

    return field[key];
  }

  return field;
};
