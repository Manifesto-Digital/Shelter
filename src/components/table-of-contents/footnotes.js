import React from 'react';
import PropTypes from 'prop-types';
import { Wrapper, ReferenceList, FootNotesTitle } from './styles';

export const TableOfContentsFootNotes = ({
  referenceList,
  removeBottomMargin,
}) => (
  <Wrapper removeBottomMargin={removeBottomMargin}>
    <FootNotesTitle>Footnotes</FootNotesTitle>
    <ReferenceList>
      {referenceList.map((reference, i) => (
        <li key={i}>
          <a
            id={`reference-${i + 1}`}
            href={`#source-${i + 1}`}
            title={reference}
          >
            {`[${i + 1}] `}
          </a>
          <p>{reference.replace(/[[\]']+/g, '')}</p>
        </li>
      ))}
    </ReferenceList>
  </Wrapper>
);

TableOfContentsFootNotes.propTypes = {
  referenceList: PropTypes.array,
  removeBottomMargin: PropTypes.bool,
};

export default TableOfContentsFootNotes;
