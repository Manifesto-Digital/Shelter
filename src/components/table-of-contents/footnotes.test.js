import React from 'react';
import { TableOfContentsFootNotes } from './footnotes';
import { snapshotComponent } from '../../../__tests__/helpers/index';

export const createReferenceList = () => [
  '[s.3 Public Order Act 1986.]',
  '[s.2 Public Order Act 1986.]',
  '[s.2 Public Order Act 1986.]',
  '[s.4A(1) Public Order Act 1986.]',
  '[s.24 Police and Criminal Evidence Act 1984]',
  '[s.25 Police and Criminal Evidence Act 1984]',
  '[s.1(3) Police and Criminal Evidence Act 1984.]',
];

it('renders correctly', () => {
  const mockData = createReferenceList();
  snapshotComponent(<TableOfContentsFootNotes referenceList={mockData} />);
});
