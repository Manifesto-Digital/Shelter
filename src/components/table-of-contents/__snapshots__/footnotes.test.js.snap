// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  margin-bottom: 40px;
}

.c2 {
  padding-left: 0;
  margin-left: 0;
  list-style: none;
  font-size: 0.775rem;
}

.c2 a {
  margin-right: 10px;
}

.c2 li {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 p {
  margin-bottom: 10px;
}

.c1 {
  margin-bottom: 30px;
}

@media (min-width:62em) {
  .c2 {
    margin-left: -25px;
  }
}

<div
  className="c0"
>
  <h3
    className="c1"
  >
    Footnotes
  </h3>
  <ul
    className="c2"
  >
    <li>
      <a
        href="#source-1"
        id="reference-1"
        title="[s.3 Public Order Act 1986.]"
      >
        [1] 
      </a>
      <p>
        s.3 Public Order Act 1986.
      </p>
    </li>
    <li>
      <a
        href="#source-2"
        id="reference-2"
        title="[s.2 Public Order Act 1986.]"
      >
        [2] 
      </a>
      <p>
        s.2 Public Order Act 1986.
      </p>
    </li>
    <li>
      <a
        href="#source-3"
        id="reference-3"
        title="[s.2 Public Order Act 1986.]"
      >
        [3] 
      </a>
      <p>
        s.2 Public Order Act 1986.
      </p>
    </li>
    <li>
      <a
        href="#source-4"
        id="reference-4"
        title="[s.4A(1) Public Order Act 1986.]"
      >
        [4] 
      </a>
      <p>
        s.4A(1) Public Order Act 1986.
      </p>
    </li>
    <li>
      <a
        href="#source-5"
        id="reference-5"
        title="[s.24 Police and Criminal Evidence Act 1984]"
      >
        [5] 
      </a>
      <p>
        s.24 Police and Criminal Evidence Act 1984
      </p>
    </li>
    <li>
      <a
        href="#source-6"
        id="reference-6"
        title="[s.25 Police and Criminal Evidence Act 1984]"
      >
        [6] 
      </a>
      <p>
        s.25 Police and Criminal Evidence Act 1984
      </p>
    </li>
    <li>
      <a
        href="#source-7"
        id="reference-7"
        title="[s.1(3) Police and Criminal Evidence Act 1984.]"
      >
        [7] 
      </a>
      <p>
        s.1(3) Police and Criminal Evidence Act 1984.
      </p>
    </li>
  </ul>
</div>
`;
